File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/datasource/DataSourcesAreBackedByInterfaces.kt
package com.perrystreet.woof.konsist.datasource

import com.lemonappdev.konsist.api.ext.list.withNameEndingWith
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class DataSourcesAreBackedByInterfaces : BehaviorSpec() {
    init {
        Given("A data source interface") {
            val interfaces = KonsistUtils.dataSourceModule.interfaces().withNameEndingWith("DataSource")

            Then("Its name starts with I") {
                interfaces.assertTrue(message = Message) { it.name.startsWith("I") }
            }
        }

        Given("A data source implementation") {
            val implementations = KonsistUtils.dataSourceModule.classes().withNameEndingWith("DataSource")

            Then("It implements an I*DataSource interface") {
                implementations.assertTrue(message = Message) { implementation ->
                    implementation.hasParent { parent -> parent.name.startsWith("I") && parent.name.endsWith("DataSource") }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Every data source is an I*DataSource interface with local and fake implementations.",
            why = "The interface is the seam between the app and the outside world. Repositories, tests, and a future remote implementation all program against it.",
            howToFix = "Declare interface IFooDataSource, then make FooLocalDataSource and FakeFooDataSource implement it.",
            badExample = "class DogsLocalDataSource { fun getDogs(): Single<DogsPageDTO> }",
            goodExample = "interface IDogsDataSource { fun getDogs(): Single<DogsPageDTO> }\nclass DogsLocalDataSource : IDogsDataSource",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DataSource/DataSourcesAreBackedByInterfaces.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class DataSourcesAreBackedByInterfaces: QuickSpec {
    override class func spec() {
        Given("A data source protocol") {
            let protocols = WoofHarmonize.dataSourcePackage.protocols().withNameContaining("DataSource")

            Then("Its name ends with DataSourceImplementing") {
                protocols.assertTrue(message: message) { $0.name.hasSuffix("DataSourceImplementing") }
            }
        }

        Given("A data source implementation") {
            let implementations = WoofHarmonize.dataSourcePackage.classes().withSuffix("DataSource")

            Then("It conforms to a *DataSourceImplementing protocol") {
                implementations.assertTrue(message: message) { implementation in
                    implementation.inheritanceTypesNames.contains { $0.hasSuffix("DataSourceImplementing") }
                }
            }
        }
    }

    private static let message = LintRuleMessage(
        rule: "Every data source is a *DataSourceImplementing protocol with local and fake implementations.",
        why: "The protocol is the seam between the app and the outside world. Repositories, tests, and a future remote implementation all program against it.",
        howToFix: "Declare protocol FooDataSourceImplementing, then make FooLocalDataSource and FakeFooDataSource conform to it.",
        badExample: "final class DogsLocalDataSource { func getDogs() -> AnyPublisher<DogsPageDTO, DataSourceError> }",
        goodExample: "protocol DogsDataSourceImplementing { func getDogs() -> AnyPublisher<DogsPageDTO, DataSourceError> }\nfinal class DogsLocalDataSource: DogsDataSourceImplementing"
    )
}
